test/unit/ui/map.test.js

```diff
@@ -1466,6 +1466,281 @@ test('Map', (t) => {
+            const map = createMap(t, {
+                style: {
+                    "version": 8,
+                    "sources": {
+                        "geojson": createStyleSource()
+                    },
+                    "layers": []
+                }
+            });
+            map.on('load', () => {
+                map.setFeatureState({ source: 'geojson', id: 12345}, {'hover': true});
+                map.removeFeatureState({ source: 'geojson', id: 12345}, 'hover');
+                const fState = map.getFeatureState({ source: 'geojson', id: 12345});
+                t.equal(fState.hover, undefined);
+                t.end();
+            });
+        });
+        t.test('remove all state properties of one feature', (t) => {
+            const map = createMap(t, {
+                style: {
+                    "version": 8,
+                    "sources": {
+                        "geojson": createStyleSource()
+                    },
+                    "layers": []
+                }
+            });
+            map.on('load', () => {
+                map.setFeatureState({ source: 'geojson', id: 1}, {'hover': true, 'foo': true});
+                map.removeFeatureState({ source: 'geojson', id: 1});
+
+                const fState = map.getFeatureState({ source: 'geojson', id: 1});
+                t.equal(fState.hover, undefined);
+                t.equal(fState.foo, undefined);
+
+                t.end();
+            });
+        });
+        t.test('other properties persist when removing specific property', (t) => {
+            const map = createMap(t, {
+                style: {
+                    "version": 8,
+                    "sources": {
+                        "geojson": createStyleSource()
+                    },
+                    "layers": []
+                }
+            });
+            map.on('load', () => {
+                map.setFeatureState({ source: 'geojson', id: 1}, {'hover': true, 'foo': true});
+                map.removeFeatureState({ source: 'geojson', id: 1}, 'hover');
+
+                const fState = map.getFeatureState({ source: 'geojson', id: 1});
+                t.equal(fState.foo, true);
+
+                t.end();
+            });
+        });
+        t.test('remove all state properties of all features in source', (t) => {
+            const map = createMap(t, {
+                style: {
+                    "version": 8,
+                    "sources": {
+                        "geojson": createStyleSource()
+                    },
+                    "layers": []
+                }
+            });
+            map.on('load', () => {
+                map.setFeatureState({ source: 'geojson', id: 1}, {'hover': true, 'foo': true});
+                map.setFeatureState({ source: 'geojson', id: 2}, {'hover': true, 'foo': true});
+
+                map.removeFeatureState({ source: 'geojson'});
+
+                const fState1 = map.getFeatureState({ source: 'geojson', id: 1});
```

Comment:
After clearing all the state in the source/source layer, there should be no feature objects remaining. Ideally `fState1` returns `null`/`undefined`.

Reply:
@asheemmamoowala `getFeatureState()` returns `{ }` before any state gets set, and I think it makes sense to return that after add+remove, for the sake of API symmetry